tests/testthat/test-metareactive.R

```diff
@@ -37,12 +37,106 @@ test_that("doesn't break metaprogramming with quosures", {
 
     expect_identical(result1, "ok")
 
-    expect_snapshot_output(formatCode(withMetaMode(o1())))
-
-    # Outputs
+    expect_snapshot_output(withMetaMode(o1()))
+  })
 
+  # Outputs
+  testthat::skip_if_not_installed("shiny", "1.6.0.9000")
+  isolate({
     out1 <- rlang::inject(metaRender(shiny::renderText, !!outer_quo))
     expect_identical(out1(), "ok")
+  })
+})
+
+describe("metaAction", {
+  it("basically works", {
+    a <- 1
+    metaAction({
+      a <- 2
+    })
+    expect_identical(a, 2)
+
+    metaAction(quote({
+      a <- 3
+    }), quoted = TRUE)
+    expect_identical(a, 3)
+
+    env <- new.env()
+    metaAction({
+      a <- 4
+    }, env = env)
+    expect_identical(a, 3)
+    expect_identical(env[["a"]], 4)
+
+    metaAction(quote({
+      a <- 5
+    }), env = env, quoted = TRUE)
+    expect_identical(a, 3)
+    expect_identical(env[["a"]], 5)
+  })
+
+  it("unquotes properly", {
+    b <- TRUE
+    act <- metaAction(x <- ..(b))
+    expect_identical(x, TRUE)
+    expect_snapshot_output(withMetaMode(act()))
+
+    mr <- metaReactive({
+      FALSE
+    })
+    isolate({
+      act <- metaAction(y <- ..(mr()))
+    })
+    expect_identical(y, FALSE)
+    expect_snapshot_output(expandChain(act()))
+  })
+
+  it("errors on non-meta usage", {
+    ma <- metaAction({})
+    expect_error(ma())
+  })
+
+  it("can contain code that uses !!", {
+    ma <- metaAction({
+      foo <- 1
+      x <- rlang::expr(!!foo)
+    })
+    expect_identical(x, 1)
+    expect_snapshot_output(withMetaMode(ma()))
```

Comment:
Turns out on R3.4 and lower you'll get

```r
> quote({a <- 1; rlang::expr(!!a + !!a)})
{
    a <- 1
    rlang::expr(!(!a + (!(!a))))
}
```

So I think we should skip this one on R 3.4 and lower